pyteal/ast/abi/util_test.py

```diff
@@ -772,3 +772,250 @@ def test_sdk_type_specs_from_signature(sig_str, sig_args, sig_rets):
+        [abi.type_spec_from_annotation(NamedTComp1)],
+    ),
+] + [
+    SafeAssignment(spec, [abi.TransactionTypeSpec()])
+    for spec in abi.TransactionTypeSpecs
+    if spec != abi.TransactionTypeSpec()
+]
+
+
+@pytest.mark.parametrize("tc", SAFE_ASSIGNMENT_TEST_CASES)
+def test_type_spec_is_assignable_safe_assignment(tc: SafeAssignment):
+    assert len(tc.bs) > 0
+    for b in tc.bs:
+        assert abi.type_spec_is_assignable_to(tc.a, b)
+        assert not abi.type_spec_is_assignable_to(b, tc.a)
+
+
+@pytest.mark.parametrize("ts", bfs_on_inheritance(abi.TypeSpec))
+def test_type_spec_is_assignable_safe_assignment_full_coverage(ts: type):
+    def exists_in_safe_assignment(_ts: type):
+        for safe_assignment in SAFE_ASSIGNMENT_TEST_CASES:
+            if type(safe_assignment.a) == _ts:
+                return True
+            for t in safe_assignment.bs:
+                if type(t) == _ts:
+                    return True
+        return False
+
+    if isabstract(ts) or all(isabstract(parent) for parent in ts.__bases__):
+        return
+    assert exists_in_safe_assignment(ts)
+
+
+class UnsafeBidirectional(NamedTuple):
+    xs: list[abi.TypeSpec]
+
+
+UNSAFE_BIDIRECTIONAL_TEST_CASES: list[UnsafeBidirectional] = [
+    UnsafeBidirectional(
+        [abi.DynamicArrayTypeSpec(abi.Uint8TypeSpec()), abi.DynamicBytesTypeSpec()]
+    ),
+    UnsafeBidirectional(
+        [abi.StaticBytesTypeSpec(7), abi.StaticArrayTypeSpec(abi.ByteTypeSpec(), 11)]
+    ),
+    UnsafeBidirectional(
+        [
+            abi.StringTypeSpec(),
+            abi.DynamicArrayTypeSpec(abi.Uint32TypeSpec()),
+            abi.DynamicArrayTypeSpec(abi.Uint64TypeSpec()),
+            abi.AddressTypeSpec(),
+            abi.StaticBytesTypeSpec(33),
+        ]
+    ),
+    UnsafeBidirectional(
+        [
+            abi.type_spec_from_annotation(NamedTDecl),
+            abi.type_spec_from_annotation(NamedTComp0),
+            abi.type_spec_from_annotation(NamedTComp2),
+        ]
+    ),
+    UnsafeBidirectional(
+        [
+            abi.type_spec_from_annotation(NamedTComp1),
+            abi.TupleTypeSpec(
+                abi.AddressTypeSpec(),
+                abi.StaticArrayTypeSpec(abi.Uint16TypeSpec(), 100),
+            ),
+        ]
+    ),
+    UnsafeBidirectional(
+        [spec for spec in abi.TransactionTypeSpecs if spec != abi.TransactionTypeSpec()]
+        + [
+            spec()
+            for spec in bfs_on_inheritance(abi.UintTypeSpec)
+            if not isabstract(spec)
+        ]
+        + [abi.BoolTypeSpec()]
+    ),
+    UnsafeBidirectional(abi.ReferenceTypeSpecs + [abi.TransactionTypeSpec()]),
+]
+
+
+@pytest.mark.parametrize("tc", UNSAFE_BIDIRECTIONAL_TEST_CASES)
+def test_type_spec_is_assignable_unsafe_bidirectional(tc: UnsafeBidirectional):
+    for ia, a in enumerate(tc.xs):
+        for ib, b in enumerate(tc.xs):
+            if ia == ib:
+                continue
+            assert not abi.type_spec_is_assignable_to(a, b)
```

Comment:
@ahangsu Optional:  As an additional sanity check, rather than omitting the identity check, I see no harm in including it.  What do you think?

If you feel the change reduces clarity and/or you prefer as is, feel welcomed to resolve the thread without change.

```suggestion
            if ia == ib:
                assert abi.type_spec_is_assignable_to(a, b)
            else:
                assert not abi.type_spec_is_assignable_to(a, b)
```

Reply:
I don't see the harm of including it, but so long as typespecs are identical, they should be bidirectional assignable. I would step even forward for change
```suggestion
            if ia == ib:
                assert abi.type_spec_is_assignable_to(a, b)
                assert abi.type_spec_is_assignable_to(b, a)
                continue
            assert not abi.type_spec_is_assignable_to(a, b)
```
What do you think?

Reply:
Makes sense to me - thanks for bringing in the change.